### ISSUE TITLE
Why is the css/scss reliant on classes?

### ISSUE THREAD

**Issue by ajdaniel** [carbon-design-system/carbon#56]
## Overview
I am using carbon-components for a project. I find the CSS really over-verbose and unweildly. Let's look at this example of a transactional modal:

```<div data-modal id="modal" class="bx--modal" tabindex="-1">
  <div class="bx--modal-container">
    <div class="bx--modal-header">
      <h4 class="bx--modal-header__label">Label (Optional)</h4>
      <h2 class="bx--modal-header__heading">Modal Title</h2>
      <button class="bx--modal-close" type="button" data-modal-close>
        <svg class="bx--modal-close__icon">
          <use xlink:href="/carbon-icons/bluemix-icons.svg#icon--close"></use>
        </svg>
      </button>
    </div>

    <div class="bx--modal-content">
      <p>Modal text description</p>
    </div>

    <div class="bx--modal-footer">
      <button class="bx--btn bx--btn--secondary" type="button" data-modal-close>Cancel</button>
      <button class="bx--btn bx--btn--primary" type="button" data-modal-primary-focus>Save</button>
    </div>
  </div>
</div>
```

There is a class for almost every element, when an element can easily be inferred by the structure. For example, instead of have three nested classes: `bx--modal` > `bx--modal--container` > `bx--modal-header`, there could be the one class on the top level element, and use `.bx--modal > div` to select the modal-container and a `<header>` element selected by `.bx--modal header`. *An overly verbose css class system is very unnecessary and clutters the HTML making it difficult to read.*

The button is another example of over-verbose and redundant information:

```<button class="bx--btn bx--btn--primary" type="button" disabled>Primary button</button>```

We know it is a button element, so why do we need a `bx-btn` class? Perhaps if we want to make our button _bluemix-ey_ just make a `bx` class for the button. And look, we have another class with _btn_ in it... `bx--btn--primary`! This makes it very difficult for a developer to remember, and to see clearly what is happening in a big html template. The above can be easily achievable like so:

```<button class="bx primary" type="button" disabled>Primary button</button>```

using these CSS selectors: `button.bx`, `button.bx.primary`.

I know all of this style is throughout the whole application. But I think there is a lot of mileage on utilising CSS more effectively by removing the amount of classes *needed* to implement the components, and rely more on structure and selectors that can work on multiple classes.

```gherkin
Feature: cleaner HTML needed
  As a developer
  I want fewer CSS classes to use in HTML
  So that I can developer faster, and have more maintainable code
```

**Comment by ajdaniel** [comment]
I think a big change like this would need a major version change too, it would break projects already using this system.

**Comment by chrisdhanaraj** [comment]
Hey Andrew! Let me go through your points and why we built the system the way we did. The *tl;dr* is that we built the system with our priority on being flexible, consistent, and descriptive at the cost of verboseness, and that's not going to change. We've found that users can glance at the markup _or_ the `css` and have a clear idea what each piece is doing. 

**1. Why _bx--_ on everything**
Namespacing essentially allows applications to consume Carbon in pieces - which is, as it were, the primary way applications are being converted. We in fact have done a lot of work to ensure that Carbon is flexible enough to be consumed it pieces - the global `reset` can be turned off with a `scss` variable and injected into each class that requires it, for example. Most of the users for the last year or so already had applications and were shifting over to the system. 

**2. Why `bx--btn bx--btn-primary` classes**
This architecture is called *B*lock *E*lement *M*odifier. While verbose, it clearly spells out the relationship between the elements. That way when a developer wants to use a secondary button, or perhaps wants to skin the button a certain way, it's always consistent. We also make sure not to clobber the global `button` - again, this is to let applications integrate smoothly. 

**3. Why not just do `.bx-modal > div`**
There's a couple problems with this style of css. The first is that it's not _descriptive_ of what it's actually doing. While I'll probably use this style of `css` on my personal projects, working in larger applications with multiple members this becomes more and more untenable. 

The other problem with using something like `.bx-modal > div` is that we don't know what other elements users are going to put in there (maybe an artifact of a framework they're using, maybe a modification for their use case). 

Hope that explains why we architected our system this way
